File: hard/123. Best Time to Buy and Sell Stock III/123. Best Time to Buy and Sell Stock III.go
package main

import "fmt"

func maxProfit(prices []int) int {
	n := len(prices)
	if n == 0 {
		return n
	}

	purchases := 2
	dp := make([][]int, purchases+1)
	for i := range dp {
		dp[i] = make([]int, n)
	}

	for purchase := 1; purchase <= purchases; purchase++ {
		minPrice := prices[0]

		for i := 1; i < n; i++ {
			minPrice = min(minPrice, prices[i]-dp[purchase-1][i-1])
			dp[purchase][i] = max(dp[purchase][i-1], prices[i]-minPrice)
		}
	}

	return dp[2][n-1]
}

func main() {
	fmt.Print(maxProfit([]int{3, 3, 5, 0, 0, 3, 1, 4}))
	fmt.Print(maxProfit([]int{1, 2, 3, 4, 5}))
	fmt.Print(maxProfit([]int{7, 6, 4, 3, 1}))
}

File: hard/123. Best Time to Buy and Sell Stock III/123. Best Time to Buy and Sell Stock III.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # buy1, buy2 = -prices[0], -prices[0]
        # profit_curr, total_profit = 0, 0

        # for price in prices:
        #     buy1 = max(buy1, -price)
        #     profit_curr = max(profit_curr, buy1 + price)
        #     buy2 = max(buy2, profit_curr - price)
        #     total_profit = max(total_profit, price + buy2)

        # return total_profit

        n = len(prices)
        if not n:
            return n

        max_purchases = 2
        dp = [[0] * n for _ in range(max_purchases + 1)]

        for k in range(1, max_purchases + 1):
            min_price = prices[0]

            for i in range(1, n + 1):
                min_price = min(min_price, prices[i] - dp[k - 1][i - 1])
                dp[k][i] = max(dp[k][i - 1], prices[i] - min_price)

        return dp[max_purchases][-1]


solution = Solution()
print(solution.maxProfit(prices=[3, 3, 5, 0, 0, 3, 1, 4]))
print(solution.maxProfit(prices=[1, 2, 3, 4, 5]))
print(solution.maxProfit(prices=[7, 6, 4, 3, 1]))
